rep-2012.rst

```diff
@@ -0,0 +1,425 @@
+
+* a request was sent by a client, but not received by a server
+* a request was received by a server, but a response was not sent
+
+Alternatively, a single event could have been defined containing both the request and response.
+While this would be convenient for tools to match requests and responses, it would result in duplicate or unused message content.
+
+A second alternative is to define unique request and response event types for clients and services (for a total of four event types and four topics per service).
+However, it's not clear that there is much benefit in the additional types considering the definition of a client request type and service request type would be identical (the same applying to response types).
+
+Define a single event type with serialized data
+-----------------------------------------------
+
+Rather than generating event types in ``rosidl``, we considered defining a single type with type-erased data for the request and/or response, for example,
+
+  .. code-block::
+
+   rcl_interfaces/msg/ServiceEventInfo info
+
+   # The request/response type
+   # e.g. my/srv/Foo_Request
+   string idl_type_name
+
+   # Serialized data
+   byte[] request_or_response
+
+This has the benefit of avoiding additional code generation for each service type and gives us the option to put all service events on one common topic.
+
+The downsides include extra overhead from serializing/deserializing the data and tools having to filter out messages based on the service type or name.
+
+Ultimately, it was decided that having separate event topics per service name would be more useful for tooling and debugging.
+For example, it makes it easier to selectively introspection a subset of services by name.
+
+
+Backwards Compatibility
+=======================
+
+The addition of service introspection should not impact existing logic.
+As an opt-in feature, users should not incur additional overhead by default.
+
+Feature Progress
+================
+
+Most elements of this proposal have been implemented and are currently under review.
+
+Progress on the implementation is being tracked on GitHub at `ros2/ros2#1285 <https://github.com/ros2/ros2/issues/1285>`_.
+
+Other
+=====
+
+
+Tooling
+-------
+
+``ros2 service``
+^^^^^^^^^^^^^^^^
+The existing ``ros2 service`` tool can be extended using an ``echo`` keyword to monitor service events.
+Internally, it would subscribe to the `hidden topics <Publishing Service Events_>`_ and echo them.
+The existing command line parameters for topics can be extended to be used with this ``echo`` verb, along with new
+arguments on to filter message content and analyze delays.
+
+Building on the example with AddTwoInts discussed earlier, an example ``ros2 service echo`` call may look like the following:
+
+.. code-block::
+
+   $ ros2 service echo /add_two_ints
+   -----------------------
+   request_type: REQUEST_SENT
```

Comment:
the actual output in the implementation is different from this design. need to update with the following correct one? or this is design, so i guess it does not need to match with current implementation?

```bash
$ ros2 service echo --flow-style /add_two_ints
info:
  event_type: REQUEST_SENT
  stamp:
    sec: 1709432402
    nanosec: 680094264
  client_gid: [1, 15, 0, 18, 86, 208, 115, 86, 0, 0, 0, 0, 0, 0, 21, 3]
  sequence_number: 247
request: []
response: []
---
info:
  event_type: REQUEST_RECEIVED
  stamp:
    sec: 1709432402
    nanosec: 680459568
  client_gid: [1, 15, 0, 18, 86, 208, 115, 86, 0, 0, 0, 0, 0, 0, 20, 4]
  sequence_number: 247
request: [{a: 2, b: 3}]
response: []
---
info:
  event_type: RESPONSE_SENT
  stamp:
    sec: 1709432402
    nanosec: 680765280
  client_gid: [1, 15, 0, 18, 86, 208, 115, 86, 0, 0, 0, 0, 0, 0, 20, 4]
  sequence_number: 247
request: []
response: [{sum: 5}]
---
info:
  event_type: RESPONSE_RECEIVED
  stamp:
    sec: 1709432402
    nanosec: 681027998
  client_gid: [1, 15, 0, 18, 86, 208, 115, 86, 0, 0, 0, 0, 0, 0, 21, 3]
  sequence_number: 247
request: []
response: []
---
...
```